rfcs/0000-go-generate.md

```diff
@@ -0,0 +1,177 @@
+require the tools specified in the `//go:generate` directives.
+
+Users can specify flags to use with the `go generate` command with the
+`BP_GO_GENERATE_FLAGS` environment variable.  The most interesting flag here is
+`-run="<regex>"` which allows users to select a subset of generate directives
+to run.
+
+Users can specify files/packages that should be generated with the
+`BP_GO_GENERATE_ARGS` environment variable.  This way, users can run generation
+on only certain files at build time.
+
+### Build behaviour
+The buildpack should do the equivalent of running
+```
+go generate <user-provided flags> <user-provided args>
+```
+from the app working directory.
+
+If the user does not provide flags, it should run
+```
+go generate ./...
+```
+from the app working directory.
+
+`go generate` fails with an informative error if a tool needed for a directive
+is not found on the `PATH`. Though the buildpack API should prevent this error
+from arising.
+
+
+### Language Family Ordering
+Proposed updated order:
+```
+[[order]]
+
+  [[order.group]]
+    id = "paketo-buildpacks/go-dist"
+
+  [[order.group]]
+    id = "paketo-buildpacks/go-generate"
+    optional = true
+
+  [[order.group]]
+    id = "paketo-buildpacks/go-mod-vendor"
+
+  [[order.group]]
+    id = "paketo-buildpacks/go-build"
+
+[[order]]
+
+  [[order.group]]
+    id = "paketo-buildpacks/go-dist"
+
+  [[order.group]]
+    id = "paketo-buildpacks/go-generate"
+    optional = true
+
+  [[order.group]]
+    id = "paketo-buildpacks/dep"
+
+  [[order.group]]
+    id = "paketo-buildpacks/dep-ensure"
+
+  [[order.group]]
+    id = "paketo-buildpacks/go-build"
+
+[[order]]
+
+  [[order.group]]
+    id = "paketo-buildpacks/go-dist"
+    version = "0.2.6"
+
+  [[order.group]]
+    id = "paketo-buildpacks/go-generate"
+    optional = true
+
+  [[order.group]]
+    id = "paketo-buildpacks/go-build"
+    version = "0.1.2"
+```
+
+Note that in some cases, `go mod vendor` can fail if run before `go generate`
+because some non-generated files depend on packages with generated files. A
+common example would be generated fakes for tests.
+
+<!---
+{{Give a high-level overview of implementation requirements and concerns. Be
+specific about areas of code that need to change, and what their potential
+effects are. Discuss which repositories and sub-components will be affected,
+and what its overall code effect might be.}}
+-->
+
+## Source Material (Optional)
+
+- An [existing go-generate buildpack](https://github.com/stefanlesperance/go-generate)
+<!---
+{{Any source material used in the creation of the RFC should be put here.}}
+-->
+
+## Unresolved Questions and Bikeshedding (Optional)
+
+- How common are the different use-cases? Which ones should we focus on addressing?
+- Does this buildpack imply the existence of a "go get" buildpack that provides
+  arbitrary go projects on the path?
+- Should `BP_GO_GENERATE` be `true` or `false` by default?
+- Is there any way layer reuse can help us?
```

Comment:
Since you can do pretty much anything in a `go:generate` directive, I'm not sure what we'd use layers for here.